atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/sweep/GetCellTimestamps.java

```diff
@@ -96,28 +95,24 @@ private void fetchBatchOfTimestampsBeginningAtStartRow() {
         byte[] rangeStart = startRowInclusive;
 
         while (timestamps.isEmpty()) {
-            Optional<byte[]> rangeEnd = determineSafeRangeEndInclusive(rangeStart);
-            if (!rangeEnd.isPresent()) {
+            List<byte[]> rows = getRows(rangeStart);
+            if (rows.isEmpty()) {
                 return;
             }
 
             // Note that both ends of this range are *inclusive*
-            List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, rangeStart, rangeEnd.get(), batchHint);
+            List<CellWithTimestamp> batch = cqlExecutor.getTimestamps(tableRef, rows, batchHint);
             timestamps.addAll(batch);
-            rangeStart = RangeRequests.nextLexicographicName(rangeEnd.get());
+            rangeStart = RangeRequests.nextLexicographicName(Iterables.getLast(rows));
```

Comment:
Shouldn't we use the last cell in the batch here, to avoid missing data?

Reply:
this seems ok to me: if the query returned zero results, then the last cell in the query could not have had any results, so no need to include it in the next one

Reply:
My thought here was that we now pass in 128 row keys (by default), but if we have wide rows then the results returned could be from the first 28 of these (for example), so by picking the 128th row key as the paging token, there's 100 row keys' worth of data that we're skipping over.

The tests are passing, though, and I'd expect this kind of case to be covered by the tests. Will do some more digging pre-merge.

Reply:
This isn't an issue - I was conflating a couple of parts of our paging logic.